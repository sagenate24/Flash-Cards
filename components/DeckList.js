import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Text, View, StyleSheet, TouchableOpacity, ScrollView } from 'react-native';
import { getDecks, getProfile } from '../utils/api';
import { receiveDecks, receiveProfile } from '../actions';
import { AppLoading } from 'expo';
import { gray, white, lightBlue, black } from '../utils/colors';

// TODO: Add TimeStamp to each deck listed

class DeckList extends Component {
  state = {
    ready: false,
  }
  componentDidMount() {
    const { dispatch } = this.props;

    getProfile()
      .then((results) => dispatch(receiveProfile(results)))

    getDecks()
      .then((results) => dispatch(receiveDecks(results)))
      .then(() => this.setState(() => ({
        ready: true,
      })))
  }

  render() {
    const { decks, profile } = this.props;
    const { ready } = this.state;

    console.log(profile)

    if (ready === false) {
      return <AppLoading />
    }

    return (
      <ScrollView style={styles.container}>
        <View>
        <TouchableOpacity style={styles.newDeck} onPress={() => this.props.navigation.navigate(
              'NewDeck'
            )}>
              <Text style={styles.btnText}>
                Create a new study deck
            </Text>
            </TouchableOpacity>
        </View>
        <Text style={styles.deckListTitle}>Your Decks</Text>
        {Object.keys(decks).map((deck) => {

          return (
            <View style={styles.item} key={deck}>
              <TouchableOpacity onPress={() => this.props.navigation.navigate(
                'Deck',
                { currentDeck: deck }
              )}>
                <Text style={{ fontSize: 20 }}>
                  {decks[deck].title}
                </Text>
                {decks[deck].questions && decks[deck].questions.length
                  ?
                  <Text style={{ fontSize: 16, color: lightBlue }}>
                    {decks[deck].questions.length > 1
                      ? decks[deck].questions.length + ` Cards`
                      : decks[deck].questions.length + ` Card`}
                  </Text>
                  :
                  <Text style={{ color: gray, fontSize: 16 }}>EMPTY</Text>
                }
              </TouchableOpacity>
            </View>
          )
        })}
      </ScrollView>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 20,
  },
  item: {
    backgroundColor: white,
    borderRadius: 2,
    padding: 20,
    marginLeft: 10,
    marginRight: 10,
    marginTop: 17,
    justifyContent: 'center',
    shadowOpacity: 0.8,
    shadowColor: 'rgba(0,0,0,0.24)',
    shadowOffset: {
      width: 0,
      height: 3,
    },
  },
  newDeck: {
    backgroundColor: lightBlue,
    borderRadius: 2,
    padding: 24,
    marginLeft: 12,
    marginRight: 12,
    marginTop: 17,
    justifyContent: 'center',
    shadowOpacity: 0.8,
    shadowColor: 'rgba(0,0,0,0.24)',
    shadowOffset: {
      width: 0,
      height: 3,
    },
  },
  deckListTitle: {
    fontSize: 20,
    color: black,
    fontWeight: 'bold',
    paddingTop: 47,
    paddingBottom: 30,
  },
  btnText: {
    color: white,
    fontSize: 22,
    textAlign: 'center',
  }
});

function mapStateToProps(decks) {
  return {
    decks,
  }
}

export default connect(mapStateToProps)(DeckList);