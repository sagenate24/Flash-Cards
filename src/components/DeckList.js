import React, { Component } from 'react';
import { connect } from 'react-redux';
import {
  Text,
  View,
  StyleSheet,
  ScrollView,
  TouchableOpacity,
} from 'react-native';
import { receiveDecks } from '../actions/decks';
import { getInitialData, profanityDetector } from '../utils/helpers';
import { receiveProfile } from '../actions/profile';
import { gray, white, black, red } from '../utils/colors';

class DeckList extends Component {
  componentDidMount() {
    const { dispatch } = this.props;

    getInitialData().then(({ decks, profile }) => {
      dispatch(receiveDecks(decks));
      dispatch(receiveProfile(profile));
    });
  }

  render() {
    const { decks, navigation, parentalControl } = this.props;

    return (
      <ScrollView style={styles.container}>
        <View>
          <TouchableOpacity
            style={styles.newDeck}
            onPress={() => navigation.navigate('NewDeck')}
          >
            <Text style={styles.btnText}>
              Create a new deck
            </Text>
          </TouchableOpacity>
        </View>
        {decks.length > 0
          ? <Text style={styles.deckListTitle}>Your Decks</Text>
          : null
        }
        <View style={{ marginBottom: 57 }}>
          {decks.length > 0
            ? Object.values(decks).map(deck => (
              <View style={styles.deck} key={deck.title}>
                <TouchableOpacity
                  onPress={() => navigation.navigate(
                    'Deck',
                    { currentDeck: deck,
                      parentalControl,
                    },
                  )}
                >
                  <Text style={{ fontSize: 20 }}>
                    {parentalControl === 'on' ? profanityDetector(deck.title) : deck.title}
                  </Text>
                  {deck.questions && deck.questions.length
                    ? (
                      <Text style={{ fontSize: 16, color: '#6ed3cf' }}>
                        {deck.questions.length > 1
                          ? `${deck.questions.length} Cards`
                          : `${deck.questions.length} Card`}
                      </Text>
                    ) : <Text style={{ color: gray, fontSize: 16 }}>EMPTY</Text>
                  }
                </TouchableOpacity>
              </View>
            )) : (
              <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
                <View style={{ borderBottomColor: gray, borderBottomWidth: 1 }}>
                  <Text style={[styles.emptyHeader]}>Welcome to FlashCards!</Text>
                </View>
                <Text style={[styles.emptyText]}>Get started by creating your own decks.</Text>
              </View>
            )
          }
        </View>
      </ScrollView>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 20,
  },
  deck: {
    backgroundColor: white,
    borderRadius: 2,
    padding: 20,
    marginLeft: 10,
    marginRight: 10,
    marginTop: 17,
    justifyContent: 'center',
    shadowOpacity: 0.8,
    shadowColor: 'rgba(0,0,0,0.24)',
    shadowOffset: {
      width: 0,
      height: 3,
    },
  },
  newDeck: {
    backgroundColor: red,
    borderRadius: 2,
    padding: 24,
    marginLeft: 12,
    marginRight: 12,
    marginTop: 17,
    justifyContent: 'center',
    shadowOpacity: 0.8,
    shadowColor: 'rgba(0,0,0,0.24)',
    shadowOffset: {
      width: 0,
      height: 3,
    },
  },
  deckListTitle: {
    fontSize: 20,
    color: black,
    fontWeight: 'bold',
    paddingTop: 47,
    paddingBottom: 30,
  },
  btnText: {
    color: white,
    fontSize: 22,
    textAlign: 'center',
  },
  emptyHeader: {
    fontWeight: 'bold',
    color: black,
    fontSize: 23,
    marginTop: 100,
    textAlign: 'center',
  },
  emptyText: {
    opacity: 0.6,
    fontWeight: 'bold',
    fontSize: 17,
    marginTop: 20,
    textAlign: 'center',
  },
  remove: {
    alignSelf: 'flex-end',
    fontSize: 16,
    color: red,
  },
});

function mapStateToProps({ decks, profile }) {
  const deckValuesArr = Object.values(decks);
  const newDeckArr = [];

  for (let i = 0; i < deckValuesArr.length; i++) {
    if (deckValuesArr[i] !== null) {
      newDeckArr.push(deckValuesArr[i]);
    }
  }

  return {
    parentalControl: profile.parentControl,
    decks: Object.values(newDeckArr).map(deck => ({
      ...deck,
    })).sort((a, b) => b.timeStamp - a.timeStamp),
  };
}

export default connect(mapStateToProps)(DeckList);
