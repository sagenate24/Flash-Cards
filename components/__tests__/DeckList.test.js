import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';

import DeckList from '../DeckList';

const mockDecks = {
  React: {
    title: 'React',
    timeStamp: 1534284894237,
    recentScore: 35,
    questions: [
      {
        question: 'What is ReactWhat is ReactWhat is ReactWhat is ReactWhat is ReactWhat is ReactWhat is ReactWhat is ReactWhat is ReactWhat is ReactWhat is ReactWhat is ReactWhat is React?',
        answer: 'A library for managing user interfacesA library for managing user interfacesA library for managing user interfacesA library for managing user interfacesA library for managing user interfacesA library for managing user interfacesA library for managing user interfacesA library for managing user interfacesA library for managing user interfacesA library for managing user interfacesA library for managing user interfacesA library for managing user interfacesA library for managing user interfacesA library for managing user interfacesA library for managing user interfacesA library for managing user interfacesA library for managing user interfacesA library for managing user interfacesA library for managing user interfacesA library for managing user interfacesA library for managing user interfacesA library for managing user interfaces',
      },
      {
        question: 'Where do you make Ajax requests in React?',
        answer: 'The componentDidMount lifecycle event',
      },
    ],
  },
};

it('should match Snapshot', (done) => {
  const snap = renderer.create(
    <DeckList.WrappedComponent decks={mockDecks} />,
  ).toJSON();

  expect(snap).toMatchSnapshot();

  done();
});
