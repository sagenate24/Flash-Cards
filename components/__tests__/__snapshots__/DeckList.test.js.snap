// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match Snapshot 1`] = `
<Modal
  animationType="none"
  hardwareAccelerated={false}
  transparent={false}
  visible={true}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#4e8098",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      />
    </View>
    <View>
      <Text
        accessible={true}
        allowFontScaling={true}
        collapsable={undefined}
        ellipsizeMode="tail"
        style={
          Object {
            "alignItems": "center",
            "color": "#fff",
            "fontSize": 26,
            "fontWeight": "bold",
            "opacity": 0,
            "transform": Array [
              Object {
                "scale": 0.1,
              },
            ],
          }
        }
      >
        FlashCards
      </Text>
    </View>
  </View>
</Modal>
`;
